streambed/src/crypto.rs

```diff
@@ -1,35 +1,41 @@
 //! Helpers to encrypt and decrypt things, particularly message payloads.
 
 use aes::{
-    cipher::{generic_array::GenericArray, BlockEncrypt, KeyInit},
+    cipher::{generic_array::GenericArray, BlockEncrypt},
     Aes128,
 };
 use rand::RngCore;
+use sha2::{Digest, Sha256, Sha512};
+
+use hmac::Hmac;
+type HmacSha256 = Hmac<Sha256>;
 
 const BLOCKSIZE: usize = 16;
 
+pub const KEY_SIZE: usize = 16;
 pub const SALT_SIZE: usize = 12;
 
 /// Generate a random salt for the required SALT_SIZE.
-pub fn salt<T>(rng: &mut T) -> Vec<u8>
+pub fn salt<T>(rng: &mut T) -> [u8; SALT_SIZE]
 where
     T: RngCore,
 {
-    let mut bytes = vec![0; SALT_SIZE];
+    let mut bytes = [0; SALT_SIZE];
     rng.fill_bytes(&mut bytes);
     bytes
 }
 
 /// Performs an AES-128 ECB decryption using a salt and a 16 byte key
 #[inline]
-pub fn decrypt(bytes: &mut [u8], key: &[u8], salt: &[u8]) {
+pub fn decrypt(bytes: &mut [u8], key: &[u8; KEY_SIZE], salt: &[u8; SALT_SIZE]) {
     encrypt(bytes, key, salt)
 }
 
 /// Performs an AES-128 ECB encryption using a salt and a 16 byte key
-pub fn encrypt(bytes: &mut [u8], key: &[u8], salt: &[u8]) {
+pub fn encrypt(bytes: &mut [u8], key: &[u8; KEY_SIZE], salt: &[u8; SALT_SIZE]) {
```

Comment:
As this is using a block cypher, what is the implementation here based on? Is it using one of the recognised block cypher modes of operation?  e.g. there's a crate here that seems to implement these: https://github.com/RustCrypto/block-modes/

I'm not sure whether the aes crate implements any of these internally - it just seems to be doing block level encryption (although I'm not clear whether the `encrypt_blocks` function, which takes multiple blocks, is doing something more)

It might be good to include some notes about which, if any, "block cypher mode of operation" this uses anyway.

Reply:
Documented. Turns out not to be ECB. :-)